src/librustc/ty/inhabitedness/mod.rs

```diff
@@ -145,8 +152,14 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
     pub fn uninhabited_from(
                 &self,
                 visited: &mut FxHashSet<(DefId, &'tcx Substs<'tcx>)>,
+                mut recursion_depth: u32,
                 tcx: TyCtxt<'a, 'gcx, 'tcx>) -> DefIdForest
     {
+        recursion_depth += 1;
+        if recursion_depth >= ARBITRARY_RECURSION_LIMIT {
```

Comment:
We normally use the `tcx.sess.recursion_limit` and emit a fatal error when it is exceeded (grep for that for usage examples).